### PR TITLE
Retire out-of-module KEM folder

### PR DESCRIPTION
**Pull request by torben-hansen**
### Issues:

Related to CryptoAlg-2615

### Description of changes: 

PR 2 out of 3, following https://github.com/aws/aws-lc/pull/1828.

This PR retires the out-of-module KEM folder. The legacy Kyber kems are still used in TLS implementations. So, we move these to the Kyber kem definitions under the Kyber folder. After the re-organisation, there was redundant glue code. For example:
```
kyber768r3_keygen --> kyber512r3_keypair --> pqcrystals_kyber768_ref_keypair
```
where `pqcrystals_kyber768_ref_keypair` was the actual implementation. `kyber512r3_keypair` was used as an intermediate to transform the return code from the implementation, since it had reverted the usual success/failure return codes. But we can do that transform in `kyber768r3_keygen`.

I removed the table search method completely. Legacy KEM objects are now completely static and a reference is obtained through a function `get_legacy_kem_kyber*()`.

Had to patch up a few paths as well. Finally, moved the README to the new in-module KEM folder.

### Testing:

Existing testing ensures that we are testing availability of all the KEM implementations.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: crypto/fipsmodule/kem/kem.c [hunk 1]
```diff
@@ -169,27 +169,23 @@ DEFINE_LOCAL_DATA(KEM, KEM_ml_kem_1024) {
...
       return KEM_ml_kem_768();
     case NID_MLKEM1024:
       return KEM_ml_kem_1024();
+    // Try legacy KEMs.
+    case NID_KYBER512_R3:
+      return get_legacy_kem_kyber512_r3();
+    case NID_KYBER768_R3:
+      return get_legacy_kem_kyber768_r3();
+    case NID_KYBER1024_R3:
+      return get_legacy_kem_kyber1024_r3();
```
Comment (samuel40791765) [inline crypto/fipsmodule/kem/kem.c:185]: Minor Nit: Should we name these something like `KEM_legacy_kem_kyber512_r3` to align with the rest of our `KEM_*` naming scheme?

**Review by samuel40791765 — APPROVED** [review]

File: crypto/fipsmodule/kem/kem.c (hunk 1 above)
Comment (torben-hansen) [reply crypto/fipsmodule/kem/kem.c:185]: Will be removed at some point when can retire the now old Kyber implementation.

File: crypto/kyber/kem_kyber.c [hunk 2]
```diff
@@ -3,100 +3,188 @@
...
+// and fails (0). Define wrappers to handle that.
 
-int kyber512r3_keypair_deterministic(uint8_t *public_key  /* OUT */,
-                                     uint8_t *secret_key  /* OUT */,
-                                     const uint8_t *seed  /* IN */) {
-  return pqcrystals_kyber512_ref_keypair_derand(public_key, secret_key, seed);
+static int kyber512r3_keygen_deterministic(uint8_t *public_key,
+                                           uint8_t *secret_key,
+                                           const uint8_t *seed) {
+  return pqcrystals_kyber512_ref_keypair_derand(public_key, secret_key, seed) == 0;
```
Comment (andrewhop) [inline crypto/kyber/kem_kyber.c:17]: Where were we checking the return code of `kyber512r3_keypair_deterministic` before that you're now checking the return code of `pqcrystals_kyber512_ref_keypair_derand` here in `kyber512r3_keygen_deterministic`? Does that now need to handle this function returning our "normal" return code?

File: crypto/kyber/kem_kyber.c [hunk 3]
```diff
@@ -3,100 +3,188 @@
...
-                            const uint8_t *ciphertext /* IN  */,
-                            const uint8_t *secret_key /* IN  */) {
-  return pqcrystals_kyber1024_ref_dec(shared_secret, ciphertext, secret_key);
+static int kyber1024r3_decaps(uint8_t *shared_secret,
+                              const uint8_t *ciphertext,
+                              const uint8_t *secret_key) {
+  return pqcrystals_kyber1024_ref_dec(shared_secret, ciphertext, secret_key) == 0;
+}
+
+const KEM_METHOD kem_kyber1024r3_method = {
```
Comment (andrewhop) [inline crypto/kyber/kem_kyber.c:120]: Missing static on all of these KEM_METHODs and KEMs.

File: crypto/kyber/kem_kyber.h [hunk 4]
```diff
@@ -25,65 +27,9 @@
...
-int kyber1024r3_encapsulate(uint8_t *ciphertext       /* OUT */,
-                          uint8_t *shared_secret    /* OUT */,
-                          const uint8_t *public_key /* IN  */);
-
-int kyber1024r3_decapsulate(uint8_t *shared_secret    /* OUT */,
-                          const uint8_t *ciphertext /* IN  */,
-                          const uint8_t *secret_key /* IN  */);
+const KEM * get_legacy_kem_kyber512_r3(void);
+const KEM * get_legacy_kem_kyber768_r3(void);
+const KEM * get_legacy_kem_kyber1024_r3(void);
```
Comment (andrewhop) [inline crypto/kyber/kem_kyber.h:32]: Is there a reason for the weird spacing?
```suggestion
const KEM *get_legacy_kem_kyber512_r3(void);
const KEM *get_legacy_kem_kyber768_r3(void);
const KEM *get_legacy_kem_kyber1024_r3(void);
```

File: crypto/kyber/kem_kyber.c (hunk 2 above)
Comment (andrewhop) [reply crypto/kyber/kem_kyber.c:17]: Ahhhh found it in the deleted files.

**Review by andrewhop — APPROVED** [review]

**Pull request merged**